stan/math/rev/core/autodiffstackstorage.hpp

```diff
@@ -9,28 +9,45 @@ namespace math {
- * For performance reasons the singleton is a global static for the
- * case of no threading which is returned by a function. This design
- * should allow the compiler to apply necessary inlining to get
- * maximal performance. However, this design suffers from "the static
- * init order fiasco"[0].  Anywhere this is used, we must be
- * absolutely positive that it doesn't matter when the singleton will
- * get initialized relative to other static variables.  In exchange,
- * we get a more performant singleton pattern for the non-threading
- * case. In the threading case we use the defacto standard C++11
- * singleton pattern relying on a function wrapping a static local
- * variable. This standard pattern is expected to be well supported
- * by the major compilers (as its standard), but it does incur some
- * performance penalty.  There has been some discussion on this; see
- * [1] and [2] and the discussions those PRs link to as well.
+ * For performance reasons the singleton is a global static pointer
+ * which is stored as thread local (TLS) if STAN_THREADS is
+ * defined. The use of a pointer is motivated by performance reasons
+ * for the threading case. When a TLS is used then initialization with
+ * a constant expression is required for fast access to the TLS. As
+ * the AD storage struct is non-POD it must be initialized as a
+ * dynamic expression such that compilers will wrap any access to the
+ * TLS by a TLS wrapper function which causes a significant
+ * slow-down. A pointer to the AD storage instance can be initialized
+ * to a compile-time constant expression of nullptr. In this case the
+ * compiler avoids the use of a TLS wrapper function. Furthermore we
+ * use the __thread keyword on compilers which support it. The
+ * __thread keyword is a compiler-specific (gcc, clang, Intel)
+ * extension which requires initialization with a compile time
+ * constant expression. The C++11 keyword thread_local does allow for
+ * constant and dynamic initialization of the TLS. Thus, only the
+ * __thread keyword gurantees that constant initialization and it's
+ * implied speedup, is used.
  *
- * These are thread_local only if the user asks for it with
- * -DSTAN_THREADS. This is primarily because Apple clang compilers
- * before 2016 don't support thread_local and the additional
- * performance cost. We have proposed removing support for those[3],
- * and at that time we should evaluate the performance of a switch to
- * thread_local.  If there is no loss in performance, we can remove
- * this ifdef.
+ * The initialzation of the AD instance is handled by the init()
+ * method which gets called by the constructor of the
+ * singleton. Within stan-math the initialization of the AD instance
+ * for the main thread of the program is handled by instantiating once
+ * the singleton once in the init_chainablestack.hpp file. Whenever
+ * STAN_THREADS is defined then all created child threads must call the
```

Comment:
Is this current? I gather I'm supposed to create an instance of stan::math::ChainableStack (rather than explicitly call any `init()` method).

Reply:
That was out of date. I just pushed an update to it. Originally I thought it suffices to do an `init()`, but we also need to clean the AD stack from the memory which is why I am now using a RAII style handling for the AD tape.

To see how things work in code you may want to have a look at the respective unit test:

`math/test/unit/math/rev/core/thread_stack_instance_test.cpp`